README
NAME
    Dist::Zilla::Util::SimpleMunge - Make munging File::FromCode and
    File::InMemory easier.

VERSION
    version 1.000001

SYNOPSIS
      use Dist::Zilla::Util::SimpleMunge qw( auto_munge_file );
      ...;

      sub somesub {
        ...;

        next if $file->is_bytes;

        if ( $file->can('code') ) {

        auto_munge_file $file_from_zilla, sub {
            my ( $file, $content , $encoding ) = @_;
            return $mangled if $encoding ne 'text'; # bytes or text
            ... mangle $content here ...;
            return $mangled;
        };
      }

DESCRIPTION
  NOTE: STOPGAP
    This module is mostly a stopgap and a implementation experiment in lieu
    of something better in Dist::Zilla eventually transpiring.

  BASIC USAGE
    Munging files in Dist::Zilla can be a pain.

    Its mostly the same:

      $file->content( substr( $file->content, 0, 10 ) ); # etc.

    Except when you come to "CodeRef"s, that all changes.

      my $orig_code = $file->code();
      $file->code( sub {
          $file->$orig_code() =~ s/foo/bar/
      });

    Which quickly gets messy.

    So this module is as simple as I think I can get it without hacking
    Dist::Zilla directly.

      auto_munge_file $file, sub {
         my ( $thefile, $content, $encoding ) = @_;
      };

    The callback will be called as appropriate.

    *   $content will contain the content, *decoded if possible*

    *   $encoding will be either "text" or "bytes", the latter if decoding
        is not possible.

    *   "InMemory" will apply the code immediately

    *   "FromCode" will take your code and create a chained system so your
        code will be evaluated when the file itself is written out.

    And this is the most useful and straight forward interface that doesn't
    invoke any weird re-blessing magic.

  ADVANCED USAGE
    There are a few less simple utilities that may also prove useful.

    *   "munge_InMemory" - trusts you know what you're dealing with and
        munges an "InMemory" instance via the callback.

    *   "munge_FromCode" - trusts you when you say you have a "FromCode",
        and munges with "CodeRef" chaining.

    *   "inplace_replace" - A bit of magic to replace an object in-place
        without modifying any containers that point to it and without
        changing the reference address.

    *   "to_InMemory" - returns a "FromCode" represented as a new "InMemory"
        object.

    *   "to_FromCode" - returns an "InMemory" represented as a new
        "FromCode" object.

    *   "inplace_to_InMemory" - like "to_InMemory", but replaces the object
        in-place.

    *   "inplace_to_FromCode" - like "to_FromCode", but replaces the object
        in-place.

    *   "munge_file" - combines all of the above behaviors based on
        configuration values.

    *   "munge_files" - applies a single configuration and callback to a
        collection of files.

FUNCTIONS
  "auto_munge_file"
      # auto_munge_file ( $FILE, $CODEREF )

      auto_munge_file( $zilla_file, sub {
          my ( $file, $content, $encoding ) = @_;
          return $new_content # must still be in form $encoding
      });

  "to_InMemory"
    Given a "FromCode", return an equivalent "InMemory" file, flattening the
    callback in the process into simply a string.

      my $in_memory = to_InMemory( $from_code );

  "to_FromCode"
    Given a "InMemory" or "OnDisk", return an equivalent "FromCode" file,
    converting the content into a callback that yields that content.

      my $from_code = to_FromCode( $in_memory_or_from_disk );

  "munge_InMemory"
    Munge an "InMemory" ( or similar ) item using a callback.

      munge_InMemory( $xfile, sub {
        my ( $file, $content, $encoding ) = @_;
         ...
        return $content;
      });

    This munging is applied immediately.

  "munge_FromCode"
    Munge a "FromCode" object by replacing the "CodeRef" with a new one that
    yields the former.

      munge_FromCode( $xfile, sub {
        my ( $file, $content, $encoding ) = @_;

        $content =~ s/foo/bar/;

        return $content;
      });

    Note: this code is equivalent to:

      my $orig_code = $xfile->code;
      my $encoding  = $xfile->core_return_type;
      $xfile->code( sub {

        my $content = $xfile->$orig_code();

        $content =~ s/a/b/;

        return $content;
      });

  "inplace_replace"
    This is a rather nasty way to replace an Object in place without
    breaking references held on it.

    Consider:

      source = ADDR=0x015 = data = { x => y }
                          = class = Foo

      target = ADDR=0x017 = data = { z => a }
                          = class = Bar

      array  = ADDR=0x016 = data = [ 0x015 ]

    Then:

      delete source->{x}
      source->{z} = target->{z}
      bless source, 'Bar'

    This should result in:

      source = ADDR=0x015 = data = { z => a }
                          = class = Bar

      target = ADDR=0x017 = data = { z => a }
                          = class = Bar

      array  = ADDR=0x016 = data = [ 0x015 ]

    Yes, this is rather nasty to do this, but no good alternatives at the
    moment :).

      inplace_replace( $original_object, $replacement_object );

    This will mirror all the keys from $replacement_object to
    $original_object, and subsequently ensure $original_object is
    "reblessed" into the class of $replacement_object

  "inplace_to_FromCode"
    Shorthand for

      inplace_replace( $file, to_FromCode($file) );

  "inplace_to_InMemory"
    Shorthand for

      inplace_replace( $file, to_InMemory($file) );

  "munge_file"
      # munge_file ( $FILE , \%CONFIGURATION )

      munge_file(
        $zilla_file,
        {
          via => sub { ... },
            lazy => $laziness
        }
      );

   $FILE
    A "::Role::File" object to munge.

   %CONFIGURATION
      {
        via => $CODEREF,
        lazy => $LAZINESS,
      }

   $CODEREF
    Called to munge the file itself.

    Passed a reference to the "::Role::File" instance, and a scalar
    containing the contents of that file.

    Return new content for the file via "return"

      sub {
        my ( $file, $content ) = @_ ;
        ...;
        return $newcontent;
      }

   $LAZINESS
    Specify how lazy you want the munge to be performed. Normally, what this
    is set to is dependent on the type of file being munged.

      $LAZINESS = undef ;  # use default for the file type
      $LAZINESS = 0     ;  # Munge immediately
      $LAZINESS = 1     ;  # Defer munging till as late as possible.

    For things that are normally backed by scalar values, such as
    "::File::OnDisk" and "::File::InMemory" , the laziness is equivalent to
    " $LAZINESS = 0 ", which is not lazy at all, and munges the file content
    immediately.

    For things backed by code, such as "::File::FromCode" , munging defaults
    to " $LAZINESS = 1 ", where the actual munging sub you specify is
    executed as late as possible.

    You can specify the $LAZINESS value explicitly if you want to customize
    the behavior, i.e.: Make something that is presently a scalar type get
    munged as late as possible ( converting the file into a "FromCode" file
    ), or make something currently backed by code get munged "now", (
    converting the file into a "InMemory" file )

  munge_files
    This is mostly a convenience utility for munging a lot of files without
    having to hand-code the looping logic.

    It basically just proxies for "munge_file".

      # munge_files ( \@FILEARRAY , \%CONFIGURATION )

      munge_files( [ $zilla_file_one, $zilla_file_two, ], {
        via => sub { ... },
        lazy => $laziness,
      });

   @FILEARRAY
    An "ArrayRef" of "$FILE"

   See Also
    *   "%CONFIGURATION"

    *   "$CODEREF"

    *   "$FILE"

    *   "$LAZINESS"

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Dist-Zilla-Util-FileMunge-Simple

1.000003 2017-03-10T14:48:59Z 62ffa3d

1.000002 2017-03-10T14:45:08Z ef9039f
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: +1 ↓1 (recommends: ↑1)
 - develop: +18 ↑4 -6 (recommends: +1, suggests: +2 -1)
 - runtime: ↓1 -1
 - test: +2 ↓1 -3 (recommends: ↑2 ↓1 -1)

1.000001 2014-04-27T11:41:40Z
 [Bugfix]
 - Minor documentation bugs with formatting.

1.000000 2014-04-27T11:25:29Z
 [00 Major]
 - This is a major set of changes, but mostly forwards compatible.
 - Only a few minor packaging backcompat risks for downstream $VENDORs who work by HAND.

 [01 DZil 5 Encoding Support]
 - Dist::Zilla >= 5.0 is now mandatory
 - Callback functions now pass an "encoding" value ( either 'text' or 'bytes' )
 - This is mostly a pragmatic choice, if it can be decoded, you probably want to work with it decoded
 - Otherwise you'll probably want to work with it as bytes.

 [02 New Functions]
 - A whole bunch of new functions have been added which are more DWIM and trying smarter to advoid the horrible black
   magic. - auto_munge_file now the most recommended approach, but supports no laziness overiding.

 [03 Less Black Magic]
 - There's still a rebless trick or two to cast the refs in-place, but they should be more stable and less prone to
   initialization-based errors.

 [04 New Tooling]
 - ExtUtils::MakeMaker has taken precedence over Module::Build
 - Dependencies on latest versions are now only recommended, not required.

 [05 New Verson Scheme]
 - 0.y.z -> 1.yyyzzz
 - This is the primary reason for the 1.x on the box.

 [Dependencies::Aggregate]
 - 5.008 : 200
 - 5.010 : 183
 - 5.019 : 161

 [Dependencies::Stats]
 - Dependencies changed since 0.2.2, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +43 -2 (recommends: -1, suggests: +1 -1)
 - runtime: +3 ↑1
 - test: +5 ↓1 -2 (recommends: +4)

0.2.2 2013-05-15T15:32:05Z
 - Minor Maintenance release.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.1, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +6 (suggests: ↑1)
 - test: ↑1

 [Documentation]
 - Add README.mkdn to src tree.
 - Update Copyright Year.

 [Meta]
 - Bug tracker to github issues.

 [Packaging]
 - Use MB 0.4004 test_requires

0.2.1 2012-05-09T07:58:41Z
 - Maintenance Release.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - develop: (recommends: ↑1, suggests: ↑1)
 - runtime: +1

 [Packaging]
 - Update LICENSE ( Year, Address , Indent )

 [Tests]
 - add minimum version test.

0.2.0 2011-10-09T06:46:42Z
 [Dependencies::Stats]
 - Dependencies changed since 0.1.2, see misc/*.deps* for details
 - runtime: +2
 - test: -2

 [Features]
 - Exhaustive error reporting data and input validation. ( 0.1.2-source-3-g45a8173 )
 - Added type-coercion-in-munge support. ( 0.1.2-source-2-g23ded3 )

 [Tests]
 - Added test for type-coercion-in-munge ( 0.1.2-source-2-g23ded3 )

0.1.2 2011-10-07T18:44:02Z
 - First non-trial release now demonstratably able to do 99% of its intended task.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - test: +2

 [Tests]
 - Now tests the core munge behaviour via munge_file. ( 0.1.1-source-2-g8afd9ef )

0.1.1 2011-10-07T17:21:53Z
 [Internals]
 - Rehash native munging interface to be more flexible, optional, and extendable. ( 0.1.0-source-2-gd1911aa )

 [Packaging]
 - Try out not-yet-released patches to the Dzil Twitter Plugin for supporting -TRIAL

0.1.0 2011-10-07T05:33:55Z
 - First version, Proof of Concept, but not yet feature complete.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Make munging File::FromCode and File::InMemory easier.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Util-SimpleMunge",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Util::SimpleMunge",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dist::Zilla" => "5.000",
    "Dist::Zilla::File::FromCode" => 0,
    "Dist::Zilla::File::InMemory" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::File::OnDisk" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.000003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dist::Zilla" => "5.000",
  "Dist::Zilla::File::FromCode" => 0,
  "Dist::Zilla::File::InMemory" => 0,
  "Dist::Zilla::File::OnDisk" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Scalar::Util" => 0,
  "Sub::Exporter" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    munge munges
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');
$bundle->remove_policy('NamingConventions::ProhibitMixedCaseSubs');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.015.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4205"
  },
  "configure_requires" => {
    "Module::Build" => "0.4205"
  },
  "dist_abstract" => "Make munging File::FromCode and File::InMemory easier.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Util-SimpleMunge",
  "dist_version" => "0.002003",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Util::SimpleMunge",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Dist::Zilla::File::FromCode" => 0,
    "Dist::Zilla::File::InMemory" => 0,
    "Sub::Exporter" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::More" => "1.001003",
    "version" => 0
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.4205",
  "Test::More" => "1.001003",
  "version" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
